odl/solvers/iterative/iterative.py

```diff
@@ -404,6 +394,129 @@ def gauss_newton(op, x, rhs, niter=1, zero_seq=exp_zero_seq(2.0),
         if callback is not None:
             callback(x)
 
+
+def kaczmarz(ops, x, rhs, niter=1, omega=1, projection=None,
+             callback=None):
+    """Optimized implementation of Kaczmarz's method.
+
+    Solves the inverse problem given by the set of equations::
+
+        A_n(x) = rhs_n
+
+    This is also known as the Landweber-Kaczmarz's method, since the method
+    coincides with the Landweber method for a single operator.
+
+    Parameters
+    ----------
+    ops : sequence of `Operator`'s
+        Operators in the inverse problem. They must have a
+        `Operator.derivative` property, which returns a new operator which in
+        turn has an `Operator.adjoint` property, i.e.
+        ``op[i].derivative(x).adjoint`` must be well-defined for ``x`` in the
+        operator domain and for all ``i``.
+    x : ``op.domain`` element
+        Element to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    rhs : sequence of ``ops[i].range`` elements
+        Right-hand side of the equation defining the inverse problem.
+    niter : int, optional
+        Maximum number of iterations
+    omega : positive float or sequence of positive float, optional
+        Relaxation parameter in the iteration. If a single float is given the
+        same step is used for all operators, otherwise separate steps are used.
+    projection : callable, optional
+        Function that can be used to modify the iterates in each iteration,
+        for example enforcing positivity. The function should take one
+        argument and modify it in-place.
+    callback : callable, optional
+        Object executing code per iteration, e.g. plotting each iterate.
+
+    Notes
+    -----
+    This method calculates an approximate least-squares solution of
+    the inverse problem of the first kind
+
+    .. math::
+        \mathcal{A}_i (x) = y_i \\quad 1 \\leq i \\leq n,
+
+    for a given :math:`y_n \\in \mathcal{Y}_n`, i.e. an approximate
+    solution :math:`x^*` to
+
+        :math:`\min_{x\\in \mathcal{X}}
+        \\sum_{i=1}^n \| \mathcal{A}_i(x) - y_i \|_{\mathcal{Y}_i}^2`
+
+    for a (Frechet-) differentiable operator
+    :math:`\mathcal{A}: \mathcal{X} \\to \mathcal{Y}` between Hilbert
+    spaces :math:`\mathcal{X}` and :math:`\mathcal{Y}`. The method
+    starts from an initial guess :math:`x_0` and uses the
+    iteration
+
+    .. math::
+        x_{k+1} = x_k - \omega_{[i]} \ \partial \mathcal{A}_{[i]}(x)^*
+                                 (\mathcal{A}_{[i]}(x_k) - y_{[i]}),
```

Comment:
Well there are two ways of writing the iteration, and this "swallows" the outer iterate into the `[i]` syntax. I personally feel that it is more clear this way, since we only get a single index that increases, it also makes it clear when the callbacks will be called etc.

This is your second idea.

Anyway, I'll add a comment about "other" iterates.

Will fix the argument to the derivative.